### ISSUE TITLE
FullyShardedDataParallel.summon_full_params context manger hangs in on_before_optimizer_step

### ISSUE THREAD

**Issue by lendle** [Lightning-AI/lightning#17600]
### Bug description

When using fsdp, I'm trying to compute gradient norms [per the docs](https://lightning.ai/docs/pytorch/stable/debug/debugging_intermediate.html#look-out-for-exploding-gradients) in `on_before_optimizer_step`. 

Without the `summon_full_params`, only gradients on the rank 0 gpu are computed because of sharding, so I'm trying to use that. 

The [summon_full_params docs](https://pytorch.org/docs/stable/fsdp.html#torch.distributed.fsdp.FullyShardedDataParallel.summon_full_params) state that `summon_full_params` can't be used from within a forward or backwards pass, but I don't think that would matter because `on_before_optimizer_step` wouldn't be called during either forward or backward? And if it was, the context manager [should raise an error](https://github.com/pytorch/pytorch/blob/main/torch/distributed/fsdp/_unshard_param_utils.py#L278-L281) anyway.

**If there's a better/more correct way to do this** please let me know!!

```
    def on_before_optimizer_step(self, optimizer):
        # inspect (unscaled) gradients here
        if self.global_rank == 0 and self.trainer._logger_connector.should_update_logs: 
            #only compute on logging steps
            #(is there a better way to check this than the non-public self.trainer._logger_connector.should_update_logs?)

            rank_zero_info("on_before_optimizer_step") #this is logged on the 49th step (log_every_n_steps=50 default)
            with FullyShardedDataParallel.summon_full_params(
                self.trainer.strategy.model,
                rank0_only=True,
                writeback=False,
                with_grads=True
            ):
                rank_zero_info("In summon_full_params context") #this is never logged, hangs here
                grad_norm_dict = grad_norm(self.trainer.strategy.model, norm_type=2)
            
                self.log_dict(grad_norm_dict)
```

### What version are you seeing the problem on?

master

### How to reproduce the bug

```python
It's a bit tedious to set up a fully working example. If using `summon_full_params` in `on_before_optimizer_step` is the "right"/only way to do this, let me know and I will provide a full example.
```

### Error messages and logs

Nothing, `"on_before_optimizer_step"` is logged `"In summon_full_params context"` is never logged so things seem stuck initializing the `summon_full_params` ctx manager.

### Environment

<details>
  <summary>Current environment</summary>

* CUDA:
        - GPU:
                - NVIDIA A100-SXM4-40GB
                - NVIDIA A100-SXM4-40GB
        - available:         True
        - version:           11.8
* Lightning:
        - lightning-cloud:   0.5.34
        - lightning-utilities: 0.8.0
        - pytorch-lightning: 2.1.0.dev0
        - torch:             2.0.0
        - torchdata:         0.6.0+a6c4904
        - torchmetrics:      0.11.4
* Packages:
        - aiohttp:           3.7.4.post0
        - anyio:             3.6.2
        - appdirs:           1.4.4
        - arrow:             1.2.3
        - asttokens:         2.2.1
        - async-timeout:     3.0.1
        - attrs:             22.2.0
        - awscli:            1.27.123
        - backcall:          0.2.0
        - backports.functools-lru-cache: 1.6.4
        - beautifulsoup4:    4.12.2
        - bleach:            6.0.0
        - blessed:           1.20.0
        - blinker:           1.6.2
        - botocore:          1.29.123
        - brotlipy:          0.7.0
        - cachetools:        5.3.0
        - certifi:           2022.12.7
        - cffi:              1.15.1
        - chardet:           4.0.0
        - charset-normalizer: 3.1.0
        - click:             8.1.3
        - colorama:          0.4.4
        - comm:              0.1.3
        - contourpy:         1.0.7
        - croniter:          1.3.14
        - cryptography:      40.0.2
        - cycler:            0.11.0
        - dateutils:         0.6.12
        - debugpy:           1.6.7
        - decorator:         5.1.1
        - deepdiff:          6.3.0
        - defusedxml:        0.7.1
        - docker-pycreds:    0.4.0
        - docstring-parser:  0.15
        - docutils:          0.15.2
        - entrypoints:       0.4
        - executing:         1.2.0
        - fastapi:           0.88.0
        - fastjsonschema:    2.16.3
        - filelock:          3.12.0
        - fonttools:         4.39.3
        - fsspec:            2023.4.0
        - gcsfs:             2023.4.0
        - gitdb:             4.0.10
        - gitpython:         3.1.31
        - gmpy2:             2.1.2
        - google-api-core:   2.11.0
        - google-auth:       2.17.3
        - google-auth-oauthlib: 1.0.0
        - google-cloud-core: 2.3.2
        - google-cloud-secret-manager: 2.16.1
        - google-cloud-storage: 2.8.0
        - google-crc32c:     1.1.2
        - google-resumable-media: 2.5.0
        - googleapis-common-protos: 1.57.1
        - grpc-google-iam-v1: 0.12.6
        - grpcio:            1.54.0
        - grpcio-status:     1.52.0
        - h11:               0.14.0
        - idna:              3.4
        - importlib-metadata: 6.6.0
        - importlib-resources: 5.12.0
        - inquirer:          3.1.3
        - ipykernel:         6.22.0
        - ipython:           8.13.1
        - ipywidgets:        8.0.6
        - itsdangerous:      2.1.2
        - jedi:              0.18.2
        - jinja2:            3.1.2
        - jmespath:          1.0.1
        - jsonargparse:      4.21.0
        - jsonschema:        4.17.3
        - jupyter-client:    8.2.0
        - jupyter-core:      5.3.0
        - jupyterlab-pygments: 0.2.2
        - jupyterlab-widgets: 3.0.7
        - kiwisolver:        1.4.4
        - lightning-cloud:   0.5.34
        - lightning-utilities: 0.8.0
        - markdown-it-py:    2.2.0
        - markupsafe:        2.1.2
        - matplotlib:        3.7.1
        - matplotlib-inline: 0.1.6
        - mdurl:             0.1.2
        - mistune:           2.0.5
        - mpmath:            1.3.0
        - multidict:         6.0.4
        - nbclient:          0.7.4
        - nbconvert:         7.3.1
        - nbformat:          5.8.0
        - nest-asyncio:      1.5.6
        - networkx:          3.1
        - numpy:             1.24.3
        - oauthlib:          3.2.2
        - ordered-set:       4.1.0
        - packaging:         23.1
        - pandas:            2.0.1
        - pandocfilters:     1.5.0
        - parso:             0.8.3
        - pathtools:         0.1.2
        - pexpect:           4.8.0
        - pickleshare:       0.7.5
        - pillow:            9.5.0
        - pip:               23.1.2
        - pkgutil-resolve-name: 1.3.10
        - platformdirs:      3.5.0
        - pooch:             1.7.0
        - portalocker:       2.7.0
        - prompt-toolkit:    3.0.38
        - proto-plus:        1.22.2
        - protobuf:          4.21.12
        - psutil:            5.9.5
        - ptyprocess:        0.7.0
        - pure-eval:         0.2.2
        - pyarrow:           12.0.0
        - pyasn1:            0.4.8
        - pyasn1-modules:    0.2.7
        - pycparser:         2.21
        - pydantic:          1.10.7
        - pygments:          2.15.1
        - pygraphviz:        1.10
        - pyjwt:             2.6.0
        - pyopenssl:         23.1.1
        - pyparsing:         3.0.9
        - pyrsistent:        0.19.3
        - pysocks:           1.7.1
        - python-dateutil:   2.8.2
        - python-editor:     1.0.4
        - python-multipart:  0.0.6
        - pytorch-lightning: 2.1.0.dev0
        - pytz:              2023.3
        - pyu2f:             0.1.5
        - pyyaml:            5.4.1
        - pyzmq:             25.0.2
        - readchar:          4.0.5
        - requests:          2.29.0
        - requests-oauthlib: 1.3.1
        - rich:              13.3.5
        - rsa:               4.7.2
        - s3transfer:        0.6.0
        - scipy:             1.10.1
        - sentry-sdk:        1.21.1
        - session-transformers: 0.1
        - setproctitle:      1.3.2
        - setuptools:        67.7.2
        - six:               1.16.0
        - smmap:             3.0.5
        - sniffio:           1.3.0
        - soupsieve:         2.3.2.post1
        - stack-data:        0.6.2
        - starlette:         0.22.0
        - starsessions:      1.3.0
        - sympy:             1.11.1
        - tinycss2:          1.2.1
        - torch:             2.0.0
        - torchdata:         0.6.0+a6c4904
        - torchmetrics:      0.11.4
        - tornado:           6.3
        - tqdm:              4.65.0
        - traitlets:         5.9.0
        - triton:            2.0.0
        - typeshed-client:   2.3.0
        - typing-extensions: 4.5.0
        - tzdata:            2023.3
        - urllib3:           1.26.15
        - uvicorn:           0.22.0
        - wandb:             0.15.0
        - wcwidth:           0.2.6
        - webencodings:      0.5.1
        - websocket-client:  1.5.1
        - websockets:        11.0.2
        - wheel:             0.40.0
        - widgetsnbextension: 4.0.7
        - xformers:          0.0.19
        - yarl:              1.9.1
        - zipp:              3.15.0
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - ELF
        - processor:         
        - python:            3.10.10
        - release:           4.19.0-24-cloud-amd64
        - version:           #1 SMP Debian 4.19.282-1 (2023-04-29)

</details>

cc @awaelchli @carmocca

**Comment by lendle** [comment]
I've found that if I don't set `with_grads=True` in `summon_full_params`, it does not hang. But that's not helpful here since I don't have access to sharded gradients in that case.

**Comment by awaelchli** [comment]
You are calling `summon_full_params` on rank 0 only. It is not surprising to see the misbehavior and hangs. Make sure to call it properly on all processes please.

**Comment by lendle** [comment]
Ah good catch, thanks! I was thrown off by the `rank0_only` arg, thinking it should work if I only called on rank 0. But it makes sense that it would need to be called on all ranks in order to synchronize. (Though I'm surprised it worked at all when with_grads=False).

I'll submit a ticket to pytorch to suggest the update the docs.

**Comment by awaelchli** [comment]
Yes, the rank0only flag means the params get sent to rank 0, but in order for the other ranks to know that they should send their shards, all of them need to enter that call.

I agree it's easy to get it wrong and more examples in the docs should help

**Comment by lendle** [comment]
For posterity https://github.com/pytorch/pytorch/issues/101294
